### ISSUE TITLE
Convert .ps1 to .psm1

### ISSUE THREAD

**Issue by vors** [RamblingCookieMonster/Invoke-Parallel#16]
Modules is the preferred way to distribute scripts in PowerShell world.

**Comment by RamblingCookieMonster** [comment]
Agreed - any issue with a mild file re-org?

* \Invoke-Parallel\
  * \Invoke-Parallel.psm1
* \Tests\
  * \Invoke-Parallel.Tests.ps1
  * \AppVeyor-Pester.ps1 (abstracted calls to pester to test PS2 as well, example [here](https://github.com/RamblingCookieMonster/PSDiskPart/blob/master/Tests/appveyor.pester.ps1))
* \appveyor.yml
* \README.md
* \LICENSE.md

Let me know if you see any issues with this!

**Comment by vors** [comment]
Looks good

**Comment by vors** [comment]
Also, after that's done it would make sense to add .psd1 file and publish it to PsGet gallery (and maybe PowerShellGet gallery). .psd1 file would help with versioning (it's the only option to provide a version)

**Comment by RamblingCookieMonster** [comment]
Absolutely!

Running into a strange issue. Having moved to a psm1, I can't seem to find the variables declared in the pester tests that cover ImportVariable usage.

Global, Script, Local, 0, 1, and 2 all do not contain variables with name a and long. Running the same code in the pester test works in my local session:

Successful run outside of Pester:

![image](https://cloud.githubusercontent.com/assets/6377597/6485344/2070b24a-c251-11e4-8d4c-fd2985ac6568.png)

[Failed run in Pester](https://ci.appveyor.com/project/RamblingCookieMonster/invoke-parallel/build/1.0.26) (Added temporary verbose output to illustrate what Get-Variable can see)

The Pester test fails when invoked directly on my system as well (i.e. not an issue exclusive to appveyor.pester.ps1).

Will keep tinkering.

**Comment by RamblingCookieMonster** [comment]
Yeah, I can reproduce this, https://gist.github.com/RamblingCookieMonster/05552112507db8a04246

Learned something about scope and encapsulation.

I'm assuming this might be a mutually exclusive choice between moving to a module, and allowing ImportVariables to work in all scenarios?

**Comment by vors** [comment]
That's interesting. I don't think it's mutually exclusive. Need to dig into it.

**Comment by RamblingCookieMonster** [comment]
*Sheepish*... Created a branch for the changes, committed changes to master.  Heading out for caffeine.  If it would make sense let me know and I can revert master and roll this in with a merge.  

Sorry about that!

**Comment by vors** [comment]
No worries, I think it's fine to make a force push to master in this case.

**Comment by davidwallis3101** [comment]
How far did you get with this, is there a working module that can be used for executing a script block rather than dot sourcing?

**Comment by RamblingCookieMonster** [comment]
Nope, unfortunately, I've sort of let this project languish given the excellent PoshRSJob module by Boe.  There are a few other issues that I might tackle, but this is pretty low on the priority list given what (I assume) it will take.

If you have any ideas on how to implement this, or code, totally open to pull requests : D

Cheers!

**Comment by ImN1** [comment]
Import-Modue Invoke-Parallel
$filelist | Invoke-Parallel -ImportVariables -Throttle 10 -ScriptBlock {
...
----------------------------------------
Get-RunspaceData : You cannot call a method on a null-valued expression.
At C:\Users\ImN1\Documents\WindowsPowerShell\Modules\Invoke-Parallel\Invoke-Parallel.psm1:542 char:13
+             Get-RunspaceData -wait
+             ~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Write-Error], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull,Get-RunspaceData
----------------------------------------------
but
No error by ". \PATH\Invoke-Parallel.ps1"

**Comment by sheldonhull** [comment]
fyi. This module needs to be dot sourced I believe. I have never had an error while using `. Script` while you probably are running into some scope issues with trying to import. 

@RamblingCookieMonster  I will say I've used both your project and the PoshRSJob. I find your script to be vastly simpler to implement most times. You did a great job on this!

**Comment by Silex** [comment]
What is preventing this from moving forward?
